보일러 플레이트 설정

NiNinanana opened the pull request:
## 수정
- prettier 및 eslint 설정
  - prettier 규칙
    - ![image](https://github.com/f-lab-edu/task-checker/assets/74632731/a5946172-ed75-4bae-9254-72adcebb6db7)
  - eslint 규칙
    - ![image](https://github.com/f-lab-edu/task-checker/assets/74632731/14965a6d-1207-48ce-8c7a-6b48a3bedbd0)
- 폴더 구조 설정
  - src 폴더 삭제 후 app 폴더만 사용
  - app 폴더 내 (routes) 폴더 내에 각 routes를 위한 페이지 컴포넌트 작성
  - 그 외 _로 시작하는 폴더는 private 폴더
    - _assets: 이미지, 동영상, 폰트 등 프로젝트에서 사용하는 자산 모음
    - _components: 컴포넌트 모음, 내부에 폴더로 나누어 사용
    - _types: 타입스크립트의 타입에 관한 선언 모음
    - _utils: 유틸 라이브러리나 함수, 커스텀 훅 등 유틸리티로서 사용할 소스코드 모음
  - baseUrl을 ./app으로 해두어 어느 파일이든 _components 처럼 쓰고 import 가능 - 절대경로
- jotai 설정 및 셋팅
  - 기본 사용 예시를 위해서 test atom을 만들어 두 개의 페이지에서 사용
  - 더 빠른 refresh를 위하여 @swc-jotai/react-refresh를 적용
- react-query, axios 설정 및 기본 사용
  - @tanstack/eslint-plugin-query를 사용해 eslint에 적용
  - _api 폴더에 axios로 된 fetch 함수를 만들고 그 함수를 가져와 react-query를 활용하여 불러옴
  - next server를 사용해 mock api 생성 후 사용
  - axios를 customAxios로 만들어 요청 성공 시와 에러 시에 처리가 가능하도록 작성
- 404 페이지, 에러 페이지, 로딩 페이지 구현
  - 로딩 페이지는 react.suspense를 이용해 구현하므로 컴포넌트만 생성
- 환경변수 파일을 분리하여 로컬과 배포 시의 환경변수를 따로 사용하도록 구현

## 참고
- app router의 폴더 구조 설정
  - https://nextjs.org/docs/app/building-your-application/routing/colocation
- 404 페이지
  - https://nextjs.org/docs/app/api-reference/file-conventions/not-found

NiNinanana committed:
prettier 및 eslint 설정

NiNinanana committed:
폴더 구조 설정

BuildTheTruth commented:
디렉토리 구조는 자유이긴 하나 ./src 디렉토리를 app으로 하신 이유가 있을까요?
next는 프레임워크 느낌이라서 되도록 가이드하고 있는 구조(자동으로 생성되는 구조)로 가져가시는 것이 좋을 것 같습니다.
안 그러면 설정값을 많이 건드려야 합니다.
각 디렉토리마다 _를 붙인 것도 크게 의미 없어보입니다. 만약 외부모듈과 분리를 위한 것 이라면 경로 prefix를 설정하면 될 것 같습니다.

API 호출부분과 전역 상태 관리도 예시 코드를 추가해주시면 좋을 것 같습니다.
다른 사람들과 협업할 때 정해야 하는 부분들을 떠올리시면 어떤 것들을 bolierplate에 녹여야 할지 감이 오실 겁니다.
webpack 부분은 next.config로 설정하면 되서 필요하다면 next.config 설정을 해주시면 됩니다.

NiNinanana commented:
> 디렉토리 구조는 자유이긴 하나 ./src 디렉토리를 app으로 하신 이유가 있을까요? next는 프레임워크 느낌이라서 되도록 가이드하고 있는 구조(자동으로 생성되는 구조)로 가져가시는 것이 좋을 것 같습니다. 안 그러면 설정값을 많이 건드려야 합니다. 각 디렉토리마다 _를 붙인 것도 크게 의미 없어보입니다. 만약 외부모듈과 분리를 위한 것 이라면 경로 prefix를 설정하면 될 것 같습니다.
> 
> API 호출부분과 전역 상태 관리도 예시 코드를 추가해주시면 좋을 것 같습니다. 다른 사람들과 협업할 때 정해야 하는 부분들을 떠올리시면 어떤 것들을 bolierplate에 녹여야 할지 감이 오실 겁니다. webpack 부분은 next.config로 설정하면 되서 필요하다면 next.config 설정을 해주시면 됩니다.

next.js에서 이번에 recommended로 선정된 App Router를 사용해봤습니다. src는 꼭 넣을 필요가 없고 app 폴더 내에서 page.tsx를 쓰면 해당하는 부분이 라우팅이 된다고 합니다.
디렉토리에 _를 붙인 이유는 private folder를 생성하기 위함이었습니다. _를 앞에 붙인 폴더 안에서는 page.tsx 파일을 만들어도 라우팅이 되지 않는다고 합니다.
따라서 _를 붙여서 사용했고 app 폴더를 사용하여 적용했습니다.

f-lab-dennis commented:
제가 뒤늦게 첨부해주신 next 문서를 봤네요. 처음 봤는데 괜찮은 구조 인 것 같습니다. 이걸 토대로 loading이나 error 처리도 boilerplate에 추가해주시면 좋을 것 같은데요?

NiNinanana committed:
jotai 설치 및 셋팅 후 기본 사용

NiNinanana committed:
react-query 및 axios를 설치 후 next server 활용하여 기본 테스트 진행

NiNinanana committed:
404 페이지, 에러 페이지, 로딩 페이지 구현

NiNinanana commented:
> 제가 뒤늦게 첨부해주신 next 문서를 봤네요. 처음 봤는데 괜찮은 구조 인 것 같습니다. 이걸 토대로 loading이나 error 처리도 boilerplate에 추가해주시면 좋을 것 같은데요?

- 전역 상태 관리 라이브러리는 jotai를 사용해서 설정을 해봤습니다.
- API 호출 부분은 axios와 react-query를 사용했고 next server를 활용하여 적용해봤습니다.
- 404 페이지, error 페이지, loading 페이지를 구현했습니다.

NiNinanana committed:
react-query 사용 방법 수정 및 crud 테스트
custom axios 만들어 에러 시와 성공 시 처리
각 상황에 맞추어 사용할 환경변수 분리
react-query queryclient 기본 옵션 설정

NiNinanana commented:
@f-lab-dennis
- axios를 customAxios로 따로 빼서 baseURL을 설정해주고 요청 성공 시와 실패 시 처리가 가능하도록 했습니다.
- CRUD를 다 만들어봤습니다.
- 환경변수를 두 파일로 구분하여 로컬에서는 .env.development가, 배포된 상태에서는 .env.production이 사용되도록 했습니다.
- react-query의 defaultOptions를 추가해줬습니다.
